System/AbstractFilePath/Common.hs

```diff
@@ -0,0 +1,855 @@
+  , toPlatformString
+  , toPlatformStringIO
+  , bsToPlatformString
+  , pstr
+  , packPlatformString
+#else
+  , toAbstractFilePath
+  , toAbstractFilePathIO
+  , bsToAFP
+  , afp
+  , packAFP
+#endif
+
+  -- * String deconstruction
+#if defined(WINDOWS) || defined(POSIX)
+  , fromPlatformString
+  , fromPlatformStringIO
+  , unpackPlatformString
+#else
+  , fromAbstractFilePath
+  , fromAbstractFilePathIO
+  , unpackAFP
+#endif
+
+  -- * Word construction
+  , unsafeFromChar
+
+  -- * Word deconstruction
+  , toChar
+
+  -- * Separator predicates
+  , pathSeparator
+  , pathSeparators
+  , isPathSeparator
+  , searchPathSeparator
+  , isSearchPathSeparator
+  , extSeparator
+  , isExtSeparator
+
+  -- * $PATH methods
+  , splitSearchPath,
+
+  -- * Extension functions
+    splitExtension,
+    takeExtension, replaceExtension, (-<.>), dropExtension, addExtension, hasExtension, (<.>),
+    splitExtensions, dropExtensions, takeExtensions, replaceExtensions, isExtensionOf,
+    stripExtension,
+
+    -- * Filename\/directory functions
+    splitFileName,
+    takeFileName, replaceFileName, dropFileName,
+    takeBaseName, replaceBaseName,
+    takeDirectory, replaceDirectory,
+    combine, (</>),
+    splitPath, joinPath, splitDirectories,
+
+    -- * Drive functions
+    splitDrive, joinDrive,
+    takeDrive, hasDrive, dropDrive, isDrive,
+
+    -- * Trailing slash functions
+    hasTrailingPathSeparator,
+    addTrailingPathSeparator,
+    dropTrailingPathSeparator,
+
+    -- * File name manipulations
+    normalise, equalFilePath,
+    makeRelative,
+    isRelative, isAbsolute,
+    isValid, makeValid
+  )
+where
+
+
+#ifdef WINDOWS
+import System.AbstractFilePath.Types
+import System.OsString.Windows
+    ( bsToPlatformString
+    , unsafeFromChar
+    , toChar
+    , fromPlatformString
+    , fromPlatformStringIO
+    , packPlatformString
+    , pstr
+    , toPlatformString
+    , toPlatformStringIO
+    , unpackPlatformString
+    )
+import Data.Bifunctor ( bimap )
+import qualified System.AbstractFilePath.Windows.Internal as C
+import System.AbstractFilePath.Types ()
+
+#elif defined(POSIX)
+
+import System.AbstractFilePath.Types
+import System.OsString.Posix
+    ( bsToPlatformString
+    , unsafeFromChar
+    , toChar
+    , fromPlatformString
+    , fromPlatformStringIO
+    , packPlatformString
+    , pstr
+    , toPlatformString
+    , toPlatformStringIO
+    , unpackPlatformString
+    )
+import Data.Bifunctor ( bimap )
+import qualified System.AbstractFilePath.Posix.Internal as C
+import System.AbstractFilePath.Types ()
+
+#else
+
+import System.AbstractFilePath.Internal
+    ( afp
+    , bsToAFP
+    , fromAbstractFilePath
+    , fromAbstractFilePathIO
+    , packAFP
+    , toAbstractFilePath
+    , toAbstractFilePathIO
+    , unpackAFP
+    )
+import System.AbstractFilePath.Types
+    ( AbstractFilePath )
+import System.OsString
+import System.OsString.Internal.Types
+
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
```

Comment:
Not `#ifdef WINDOWS`?

Reply:
`WINDOWS` macro is only used to generate platform-specific modules (no matter the *current* platform), because we want to be able to deal with with both posix and windows specific filepath on all platforms. `defined(mingw32_HOST_OS) || defined(__MINGW32__)` is used to select the implementation for `AbstractFilePath` and `PlatformFilePath`.